src/kvstore/NebulaStore.cpp

```diff
@@ -516,36 +516,43 @@ std::shared_ptr<Part> NebulaStore::newPart(GraphSpaceID spaceId,
 }
 
 void NebulaStore::removeSpace(GraphSpaceID spaceId) {
-  folly::RWSpinLock::WriteHolder wh(&lock_);
-  if (beforeRemoveSpace_) {
-    beforeRemoveSpace_(spaceId);
+  {
+    folly::RWSpinLock::WriteHolder wh(&lock_);
+    if (beforeRemoveSpace_) {
+      beforeRemoveSpace_(spaceId);
+    }
   }
 
-  auto spaceIt = this->spaces_.find(spaceId);
-  if (spaceIt != this->spaces_.end()) {
+  auto spaceOr = space(spaceId);
+  if (ok(spaceOr)) {
+    folly::RWSpinLock::WriteHolder wh(&lock_);
+    auto spaceIt = this->spaces_.find(spaceId);
     for (auto& [partId, part] : spaceIt->second->parts_) {
       // before calling removeSpace, meta client would call removePart to remove all parts in
       // meta cache, which do not contain learners, so we remove them here
       if (part->isLearner()) {
         removePart(spaceId, partId, false);
       }
     }
-    auto& engines = spaceIt->second->engines_;
+    this->spaces_.erase(spaceIt);
+  }
+  if (ok(spaceOr)) {
```

Comment:
See  line 527 ?  Remove redundent if sentance~

Reply:
This is not redundant. A separate if branch is used to shorten the scope of the lock.

Reply:
I am just afraid that someone will refactor this.  maybe use 
```c++
  { lock gurad; code;} 
  {other code;}
```
And it's okay now.